doc/dsl/interpretation-laufzeit.md

```diff
@@ -1,9 +1,140 @@
+
+[UML: Erzeugung Komponentenprototyp](img/create_component_prototype.png){width="50%"}
+
+**Evaluierung von Ausdrücken**
+
+Um Prototypen zu erzeugen, müssen die rechtsseitigen Ausdrücke einer Eigenschaftszuweisung (z.B.
+`x_velocity: 2.0`) evaluiert werden.
+Im obigen Beispiel handelt es sich bei diesen Ausdrücken um triviale Dezimalzahlen, es könnte sich allerdings bei
+rechtsseitigen Ausdrücken auch um Funktionsaufrufe, Verweise auf globale Objekte, etc. handeln (vgl. für
+gültige Ausdrücke hierzu [Ausdrücke](sprachkonzepte.md#ausdrücke)).
+
+Der `DSLInterpreter` ist ein [Tree-Walk-Interpreter](https://craftinginterpreters.com/a-tree-walk-interpreter.html),
+läuft also über den AST und führt für jeden so besuchten Knoten Operationen aus. Für alle Knoten, die Teil eines
+Ausdrucks sein können, erzeugt der Interpreter eine `Value`-Instanz, die den Wert des besuchten Knotens enthält.
+
+Für `NumNode` wird bspw. einfach der Wert des AST-Knotens in ein `Value` verpackt:
+```java
+public Object visit(NumNode node) {
+    return new Value(BuiltInType.intType, node.getValue());
+}
+```
+
+Für einen `GameObjectDefinition`-Knoten, der Teil eines Ausdrucks ist, ist dieses Vorgehen deutlich komplexer und wird
+unter [Typinstanziierung](#typinstanziierung) genauer erläutert. Allerdings wird auch für diesen Fall ein `Value`-Objekt
+zurückgegeben.
+
+**Anmerkung:**
+
+Die im Folgenden beschriebenen Aspekte bzgl. `quest_config` als zentralem Übergabepunkt von DSL -> Dungeon sind WIP
+und können sich daher noch grundlegend ändern (siehe hierzu [Issue #195](https://github.com/Programmiermethoden/Dungeon/issues/195)).
+
+Der zentrale Übergabepunkt zwischen dem DSL Programm und dem Dungeon-Framework wird durch die
+`quest_config`-Definition gebildet. Daher sucht der `DSLInterpreter` im nächsten Schritt die erste `quest_config`-Definition
+aus einem DSL Programm heraus und evaluiert alle Eigenschaftszuweisungen. Hierdurch werden nur Objekte und Definitionen
+evaluiert, die in Eigenschaftszuweisungen dieser `quest_config`-Definition referenziert werden.
+
+Abschließend erzeugt der `DSLInterpreter` eine `QuestConfig`-Instanz und gibt diese an das Dungeon-Framework zurück.
+Die `QuestConfig`-Instanz enthält alle Informationen für das Dungeon-Framework, um ein Dungeonlevel mit spezifizierten
+Entitäten (als `game_object`-Definition) zu erzeugen.
+
+**Wann und wie interpretiert der Interpreter?**
+
+Die Evaluierung der `quest_config`-Definition ist die **einzige** Art der Interpretation, die der `DSLInterperter`
+standardmäßig durchführt.
+Es ist aktuell kein **eigenständiger** DSL-Loop oder ähnliches vorgesehen, der parallel bzw. mit dem Gameloop des Dungeons
+mitläuft und kontinuierlich Teile des DSL-Programms ausführt.
+Die weitere Tätigkeit des `DSLInterpreter`s beschränkt sich auf die Interpretation der Event-Handler DSL-Funktionen, die
+mit Entitäten verknüpft wurden (siehe dazu [Funktionsaufrufe](#funktionsaufrufe)).
```

Comment:
> Mir wird immer noch nicht wirklich klar, worin genau die "Interpretation" liegt - es sieht für mich eher so aus, als ob Du Dir die Definitionen und so zusammensuchst und damit dann Objekte für das Spiel instanziierst. Aber vielleicht würden ein einfaches und ein komplexeres Beispiel die Sache klarer machen (die Sprachkonzepte fehlen ja auch noch - oder dieser PR muss mal rebased werden).

@cagix Das hier bezieht sich auf diese Frage. Ein Beispiel füge ich noch hinzu.

Reply:
Die Sache ist, dass so ein komplexes Beispiel (wie [hier in Szenario 2](https://github.com/Programmiermethoden/Dungeon/issues/197#issuecomment-1412144121)) aktuell noch nicht realisierbar ist, weil dafür noch einiges fehlt. Das würde ich ergänzen, wenn es wirklich geht. Fühle mich seltsam dabei, wenn ich in die Doku, die ja den realisierten Zustand beschreiben soll, zu viele Anmerkungen im Sinne von "geht jetzt alles noch nicht, ist WIP" reinschreibe. Dafür haben wir ja eigentlich die Issues. Ich könnte natürlich auch in der Doku das Issue verlinken, allerdings ist das dann wieder von Github abhängig, was ja auch nicht zielführend ist.

Reply:
> Doku, die ja den realisierten Zustand beschreiben soll, zu viele Anmerkungen im Sinne von "geht jetzt alles noch nicht, ist WIP" reinschreibe.

Bin ich bei Dir. 

> Dafür haben wir ja eigentlich die Issues. Ich könnte natürlich auch in der Doku das Issue verlinken, allerdings ist das dann wieder von Github abhängig,

Die Issues haben sich als nicht hilfreich herausgestellt. Ehrlich gesagt suche ich mir einen Wolf, wenn ich was suche ... Links im Dokument auf Issues sind so eine Sache: Einerseits funktioniert das nur im Github und nirgendwo anders. Andererseits könnte man wichtige Dinge vielleicht tatsächlich mit einem Link versehen.

---

Dennoch wäre es hilfreich, eine "Vision" zu haben: Wo wollen wir hin, was ist konkret das Ziel? D.h. könntest Du Dir ein separates Dokument vorstellen, in dem dieses Ziel beschrieben wird und eben ein paar Beispiele dazu, wie das aussehen könnte?

Reply:
> Die Issues haben sich als nicht hilfreich herausgestellt. Ehrlich gesagt suche ich mir einen Wolf, wenn ich was suche ... Links im Dokument auf Issues sind so eine Sache: Einerseits funktioniert das nur im Github und nirgendwo anders. Andererseits könnte man wichtige Dinge vielleicht tatsächlich mit einem Link versehen.

Als "nicht hilfreich" für die Dokumentation für "Konzepte, Ziele & Visionen"? Da bin ich bei dir. Mir persönlich helfen die Issues aber schon, um einen Überblick über alle Dinge zu behalten, die ich noch machen muss. Dass die wichtigsten Ideen und die "Vision" nicht über zig Issues verteilt werden sollte, sehe ich aber auch...

> Dennoch wäre es hilfreich, eine "Vision" zu haben: Wo wollen wir hin, was ist konkret das Ziel? D.h. könntest Du Dir ein separates Dokument vorstellen, in dem dieses Ziel beschrieben wird und eben ein paar Beispiele dazu, wie das aussehen könnte?

Dafür würde sich ja das Wiki anbieten. Ich bin bisher noch nicht dazu gekommen, die konkreten Ziele für die DSL da festzuhalten und dann über Milestones mit den Issues zu verknüpfen. Also Ziel/Vision-Seite im Wiki -> Milestone -> Issues. Wenn das für dich besser funktioniert, um dich in den Issues zurecht zu finden, würde ich mich da jetzt als erstes dran setzen.

Reply:
> helfen die Issues aber schon, um einen Überblick über alle Dinge zu behalten, die ich noch machen muss

Genau. Dafür sind die Tickets gut, aber für nix anderes.

> Dafür würde sich ja das Wiki anbieten. Ich bin bisher noch nicht dazu gekommen, die konkreten Ziele für die DSL da festzuhalten und dann über Milestones mit den Issues zu verknüpfen. Also Ziel/Vision-Seite im Wiki -> Milestone -> Issues. 

Weiss nicht. Damit binden wir wieder noch ein "Tool"/Konzept ein. Lass uns doch bitte bei den Tickets für "offene Aufgaben" und Projekt-Dokumentation für die Dokumentation bleiben.

Ich sehe irgendwie auch noch nicht das Problem, warum man in der Doku nicht das Ziel mit definieren kann? Warum soll das denn wieder woanders hin?

Reply:
> Ich sehe irgendwie auch noch nicht das Problem, warum man in der Doku nicht das Ziel mit definieren kann? Warum soll das denn wieder woanders hin?

Das war mein letzter Kenntnisstand über unseren Workflow, da wir das ja eigentlich so beschlossen und ins [Wiki](https://github.com/Programmiermethoden/Dungeon/wiki/Workflow) geschrieben hatten...

@AMatutat ping.

Reply:
@malt-r Du vermischt gerade operative Ziele und inhaltliche Ziele. Der operative Kram geht ins Wiki, das hatten wir so beschlossen, und da will ich auch gar nichts ändern. Ich möchte nur in der Dokumentation eine Vision sehen, wo das ganze Ding inhaltlich hinsteuert.